src/compose/service-manager.coffee

```diff
@@ -0,0 +1,276 @@
+#	return _.pick(conf, [ 'RESIN_DEVICE_RESTART', 'RESIN_RESTART' ])
+
+module.exports = class ServiceManager extends EventEmitter
+	constructor: ({ @docker, @logger, @config }) ->
+		@containerHasDied = {}
+		@listening = false
+		# Volatile state of containers, indexed by containerId (or random strings if we don't have a containerId yet)
+		@volatileState = {}
+
+	killAllLegacy: =>
+		# Containers haven't been normalized (this is an updated supervisor)
+		# so we need to stop and remove them
+		@docker.getImage(constants.supervisorImage).inspect()
+		.then (supervisorImage) =>
+			Promise.map(@docker.listContainers(all: true), (container) =>
+				@docker.getImage(container.Image).inspect()
+				.then (containerImage) =>
+					if containerImage.Id != supervisorImage.Id
+						@_killContainer(container.Id, { serviceName: 'legacy', image: container.Image }, { removeContainer: true })
+			)
+
+	reportChange: (containerId, status) ->
+		if status?
+			@volatileState[containerId] ?= {}
+			_.merge(@volatileState[containerId], status)
+			@emit('change')
+		else if containerId? and @volatileState[containerId]?
+			delete @volatileState[containerId]
+			@emit('change')
+
+	reportNewStatus: (containerId, service, status) =>
+		@reportChange(containerId, _.merge({ status }, _.pick(service, [ 'imageId', 'appId', 'releaseId', 'commit' ])))
+
+	_killContainer: (containerId, service = {}, { removeContainer = true }) =>
+		@logger.logSystemEvent(logTypes.stopService, { service })
+		@reportNewStatus(containerId, service, 'Stopping') if service.imageId?
+		containerObj = @docker.getContainer(containerId)
+		containerObj.stop(t: 10)
+		.then ->
+			containerObj.remove(v: true) if removeContainer
+		.catch (err) =>
+			# Get the statusCode from the original cause and make sure statusCode its definitely a string for comparison
+			# reasons.
+			statusCode = '' + err.statusCode
+			# 304 means the container was already stopped - so we can just remove it
+			if statusCode is '304'
+				@logger.logSystemEvent(logTypes.stopServiceNoop, { service })
+				if removeContainer
+					return containerObj.remove(v: true)
+				return
+			# 404 means the container doesn't exist, precisely what we want! :D
+			if statusCode is '404'
+				@logger.logSystemEvent(logTypes.stopRemoveServiceNoop, { service })
+				return
+			throw err
+		.tap =>
+			delete @containerHasDied[containerId] if @containerHasDied[containerId]?
+		.tap =>
+			@logger.logSystemEvent(logTypes.stopServiceSuccess, { service })
+		.catch (err) =>
+			@logger.logSystemEvent(logTypes.stopServiceError, { service, error: err })
+			throw err
+		.finally =>
+			@reportChange(containerId) if service.imageId?
+
+	kill: (service, { removeContainer = true } = {}) =>
+		@_killContainer(service.containerId, service, { removeContainer })
+		.then ->
+			service.running = false
+			return service
+
+	getAllByAppId: (appId) =>
+		@getAll("io.resin.app_id=#{appId}")
+		.filter (service) ->
+			service.appId == appId
```

Comment:
Why is the extra filtering needed? Isn't the label filter enough?

Reply:
Yup, I added the label filter later and forgot to remove this. Done.